Determine file extensions on a per-package basis

jdkoren opened the pull request:
We need a way to change extensions for generated files. Some considerations:

Generators can emit additional files, so having the generator apply/change file extensions for everything isn't ideal. We can store something in the `PackageGraph` and look it up in our models, but polluting that class is a bit of a code smell--however, we do that already for the RendererFactory, so I opted to use that existing pattern for this purpose. Happy to explore alternatives.

jcollins-g commented:
I just realized something -- that with --link-to-remote we're making an implicit assumption about how other packages are rendered and that it is the same as what we're rendering locally.

That should be possible to fix by adding some data to the table that knows how to link to remote packages.  That might influence how we decide to determine the extension -- perhaps, it should be owned by the `Package` class and derived directly rather than indirectly through the renderer.

jdkoren commented:
Changed so that `Package` determines the file type and other classes defer to that. Currently it does a simple check on `documentedWhere` as we suspect the vast majority of cases to have remote docs in html format.

Also rebased.

jdkoren commented on lib/src/model/package.dart:
```diff
@@ -181,7 +181,19 @@ class Package extends LibraryContainer
+    if (package.documentedWhere == DocumentLocation.remote) {
+      return 'html';
+    }
+    return packageGraph.rendererFactory.fileTypeRenderer.fileType;
```
I'm contemplating using the `config` member to lookup the `format` instead of using a renderer here. That requires moving `format` out of `GeneratorContext` and into `DartdocOptionContext`.

jdkoren replied:
Switched to using the config member as it's a bit more direct.
